### ISSUE TITLE
Improve interaction between hierarchical project view and tag filtering

### ISSUE THREAD

**Issue by hacst** [DependencyTrack/frontend#419]
### Current Behavior

When working with a nested project structure in which a project `a` has a child `b` with a tag `mytag`:
![tree](https://user-images.githubusercontent.com/237537/219424966-0869597e-f0d6-4db2-a289-6cfa47d6c20e.png)
clicking the `mytag` link from `b`'s project page:
![project page](https://user-images.githubusercontent.com/237537/219425207-7528911d-bbab-4750-94c8-ef04186549e9.png)
navigates to the project view and ends up showing an empty list:
![tag filtered project list](https://user-images.githubusercontent.com/237537/219426056-c4bbacd8-9537-45da-8025-ff9aad9bfdd0.png)
I found this to be surprising and it took me a while to realize what happened.

The reason this behavior is that the filtering for the tag is applied at the root level only in the hierarchical view. Manually switching back to flat project view will show all projects with the tag.

### Proposed Behavior

What I would have expected to happen for the hierarchical view would be for a pruned tree to be shown with all projects having the tag being visible regardless of their depth in the tree. This would also show parents of tagged projects without the tag themselves.

A simpler alternative or intermediary solution might be to default disable (or enforce) the show flat project view option for tag queries instead of default enabling it. This would be similar to how a search enforces the flat view right now.

### Checklist

- [X] I have read and understand the [contributing guidelines](https://github.com/DependencyTrack/dependency-track/blob/master/CONTRIBUTING.md#filing-issues)
- [X] I have checked the [existing issues](https://github.com/DependencyTrack/frontend/issues) for whether this enhancement was already requested

**Comment by rkg-mm** [comment]
Yep looks like an oversight for tag filtering. We did the simpler solution to force a flat view on searching, because filtering the data on DB level turns out to be difficult (what if a child project is included but its 2-level higher parent is not?) you need a lot of logic to get the correct result thats not doable in the DB query with the current DB structure I think, and doing in code is not an option either as then you would need to do one search in DB, then have a lot of logic requesting further missing upper-level projects etc. etc. and then applying pagination on that result is a nightmare, logic and performance wise.

so I suggest doing the same for tag filtering. forcing a flat view in that case.

**Comment by hacst** [comment]
Sounds reasonable.

**Comment by Bruceliu-rs** [comment]
May I have any update on this issue after two years please? I encountered a similar issue. when I search projects by tag, if the projects found has parent project, it shows only the parent project, and hide all 52 children projects. It can reproduce via the same link: /projects?tag=mytag. Currently the workaround is to use searchText or enable the 'Show flat project view' checkbox. The logic of hierarchical view and the logic of DB query needs to be separate. The DB query logic may need to be the same regardless it is flat view or hierarchical view. Current logic seems mixing the two together.
Thanks.
